src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs

```diff
@@ -582,57 +586,69 @@ private void RestoreUncommittedInput()
             /// </summary>
             public bool Paste()
             {
-                if (!TextView.Selection.IsEmpty)
+                using (var transaction = _textUndoHistoryRegistry.GetHistory(TextView.TextBuffer).CreateTransaction(InteractiveWindowResources.Paste))
                 {
-                    if (CutOrDeleteSelection(isCut: false))
+                    // Get text from clipboard
+                    string code;
+                    string format = Evaluator.FormatClipboard();
+                    if (format != null)
                     {
-                        MoveCaretToClosestEditableBuffer();
+                        code = format;
+                    }
+                    else if (_window.InteractiveWindowClipboard.ContainsData(ClipboardFormat))
+                    {
+                        var sb = new StringBuilder();
+                        var blocks = BufferBlock.Deserialize((string)_window.InteractiveWindowClipboard.GetData(ClipboardFormat));
+                        // Paste each block separately.
```

Comment:
We're not pasting blocks as separate submissions (and it looks like we weren't previously, despite using separate calls to ```InsertCode```). The comment can be removed.